### ISSUE TITLE
Cannot debounce API calls

### ISSUE THREAD

**Issue by basslagter** [tannerlinsley/react-query#293]
I am trying to use this for a search component, like:

```
  const { status, data, error, isFetching } = useQuery(
    searchQuery && ['results', searchQuery],
    getSearchAsync,
  );
```

I would like the call to `getSearchAsync `to be debounced while the user is typing the query. Tried adding lodash.debounce like:

```
  const { status, data, error, isFetching } = useQuery(
    searchQuery && ['results', searchQuery],
    debounce(getSearchAsync, 3000),
  );
```
But since this is a function component it is not working as expected. The `getSearchAsync `method is only called multple times after the timeout expires, which I expect to be once with the latest value (searchQuery).

**Comment by tannerlinsley** [comment]
You need to use a debounce that supports promises. It needs to return a promise immediately, then debounce off of that promise.

**Comment by basslagter** [comment]
@tannerlinsley thanks for the reply. Can you give an example?

**Comment by tannerlinsley** [comment]
https://github.com/tannerlinsley/react-table/blob/master/src/publicUtils.js

**Comment by basslagter** [comment]
Wow...there is no debounce package that does this?
Or even better...can't react-query have an option for this?

**Comment by denisborovikov** [comment]
You may try to debounce `searchQuery` change instead, something like [that](https://usehooks.com/useDebounce/).

**Comment by basslagter** [comment]
@denisborovikov thanks...that works nicely!
Still...wouldn't it be nice to have this as an option in react-query?

**Comment by timkindberg** [comment]
I'd like to see this, since react-query is serializing the key it's doing all the heavy lifting there, and could then debounce the result of the key (which devs have no hook into that result).

**Comment by mporkola** [comment]
@tannerlinsley could this be reconsidered? Debouncing based on the serialized query key would be an elegant solution, and in my view a reasonable responsibility for `react-query`. I'm interested to try my hand in making a PR if it has any chance of getting merged.

**Comment by tbntdima** [comment]
For me, if I type very fast I cannot see the input. Only after I finish typing I see the text. After that, all works fine.
Here is a my code: 
```javascript
const [searchTerm, setSearchTerm] = useState('')
const debouncedSearchTerm = useDebounce(searchTerm, 500)
const { data } = useSearch({ searchTerm: debouncedSearchTerm })
```

**Comment by denisborovikov** [comment]
@tbntdima If you use `searchTerm` as your input's value prop you should see as you type. You probably use debouncedSearchTerm instead and that's why the input is denounced. Debounced value should be used for the query only.

**Comment by tbntdima** [comment]
@denisborovikov the problem is a bit different. When `searchTerm` changes, the component was re-rendering anyway or at least tried to. And because of this, I was getting lags. My eventual solution was to move the input in upper component and do not.

**Comment by kripod** [comment]
Another issue seems to arise when dealing with debouncers. Debounced `Promise` objects get stuck in a `"loading"` state within React Query and stay inside `MutationCache` forever.

**Comment by kripod** [comment]
> You need to use a debounce that supports promises. It needs to return a promise immediately, then debounce off of that promise.

I just wrote a debouncer which fulfills this purpose:

```ts
const noop = () => {};

function debounceAsync<T>(callback: () => T, msDelay: number) {
  let cancelPrev = noop;

  return () =>
    new Promise<T>((resolve, reject) => {
      cancelPrev();

      const timeoutId = setTimeout(() => {
        resolve(callback());
      }, msDelay);
      cancelPrev = () => {
        clearTimeout(timeoutId);
        reject();
      };
    });
}
```

Usage:

```tsx
import { useMutation } from "react-query";

const debouncedCallback = debounceAsync(() => {
  console.log("This message is debounced by 400ms");
}, 400);

function Component() {
  const { mutate } = useMutation((variables) => debouncedCallback());

  return <button onClick={() => mutate()}>Test</button>
}
```

**Comment by plbertrand** [comment]
I'm fairly new to React Query but I haven't figured out how to serialize mutations to the same entry. It's necessary in the case of updating an input and avoid race conditions of two update queries. Am I wrong in thinking that this should be handled by React Query? Whatever I search to that end resolves here but even debouncing has this race condition, just less likely to happen. Is there a configuration for serializing updates? How about buffering updates while there is one in progress?

**Comment by GeekEast** [comment]
I probably found something very similar to `react-query` but with `debounce` built in.

https://ahooks.js.org/hooks/use-request/debounce

**Comment by Noitidart** [comment]
> https://github.com/tannerlinsley/react-table/blob/master/src/publicUtils.js

Does anyone have reference to this doubnce util that supports promises?

**Comment by mrlubos** [comment]
@Noitidart you might want to look at [`<DebouncedInput />`](https://github.com/TanStack/table/blob/f79330e80e5efa3c3c8680a3afa4cba193dad015/examples/react/filters/src/main.tsx#L400-L427) or if you're after the original code, it's called [`useAsyncDebounce()`](https://github.com/TanStack/table/blob/d51aac11f3ddbbb3a82cbc3852e81aa7e609c51b/src/publicUtils.js#L160-L194)

**Comment by Noitidart** [comment]
Thanks @mrlubos! I actually did something very interesting. I used `initialData` to store the debounced function (simple debounce with lodash) inside the data. I then make the `queryFn` call the method inside data. I then store the returned data in a nested key in data. Funky, but works very smoothly, it respects the react-query useQuery cache.

**Comment by roccomaniscalco** [comment]
For anyone who is interested in debouncing a mutation, I hacked together this hook.  It returns a mutation with the `debouncedMutate` function which accepts `debounceMs` as an option. Apologies for no typescript.

```js
const useDebouncedMutation = (mutationFn, options) => {
  const mutation = useMutation(mutationFn, options);
  const [isDebouncing, setIsDebouncing] = useState(false);
  let timer;

  const debouncedMutate = (variables, { debounceMs, ...options }) => {
    clearTimeout(timer);
    setIsDebouncing(true);
    timer = setTimeout(() => {
      mutation.mutate(variables, options);
      setIsDebouncing(false);
    }, debounceMs);
  };

  return { isDebouncing, debouncedMutate, ...mutation };
};
```

Example use:

```js
const channelExists = useDebouncedMutation(api.doesChannelExist);
channelExists.debouncedMutate(channelName, {
  debounceMs: 500,
});
```

I'm using this hook in a chat app to check if a user inputted channel name already exists.

**Comment by lciii** [comment]
> https://github.com/tannerlinsley/react-table/blob/master/src/publicUtils.js

FYI This link doesn't work anymore

**Comment by mrlubos** [comment]
Hey @lciii, the library has gone through numerous changes over the years, so the links pointing to the then-current version no longer work. You can find the file you referenced in my earlier comment above

**Comment by SrBrahma** [comment]
Would really be cool to have a debounceMs option in useMutation. It would be a really good ready to use functionality. +1

**Comment by AndreSilva1993** [comment]
Found this thread since I have the same issue and none of the solutions worked for me since I have a list of items and when a user selects one, I fire a fetch request to get more details about said item. If I just debounced the item change, the user would think that the item is not being selected at all. So I wanted to debounce the fetch request since I want to display the loading spinner indicating that the current item is being loaded **but** not fire the request immediately. Otherwise if the user, selected the first item and then the second, I would fire two requests unnecessarily.

I tested and I reckon this works but wanted to know your thoughts!

```
import { useRef } from 'react';
import {
  QueryKey,
  useQuery,
  useQueryClient,
  UseQueryOptions,
} from '@tanstack/react-query';

export function useDebouncedQuery(
  debounceMs: number,
  { queryFn, queryKey, ...remainingUseQueryOptions }: UseQueryOptions,
) {
  const timeoutRef = useRef<number>();
  const queryClient = useQueryClient();
  const previousQueryKeyRef = useRef<QueryKey>();

  return useQuery({
    ...remainingUseQueryOptions,
    queryKey,
    queryFn: (queryFnContext) => {
      if (!queryFn) return;

      if (previousQueryKeyRef.current) {
        queryClient.cancelQueries({ queryKey: previousQueryKeyRef.current });
      }

      previousQueryKeyRef.current = queryKey;
      window.clearTimeout(timeoutRef.current);

      return new Promise((resolve) => {
        timeoutRef.current = window.setTimeout(async () => {
          const result = await queryFn(queryFnContext);

          previousQueryKeyRef.current = undefined;
          resolve(result);
        }, debounceMs);
      });
    },
  });
}

```

**Comment by vincerubinetti** [comment]
Just want to say... I know it's not a deal-breaker to use debounced versions of query key variables, but you could also say that about implementing caching, or pagination, or one of the seemingly many many other use/edge cases that React Query covers. It's kind of weird that RQ chooses to cover all that other stuff, but not this very common (as you can tell by all the 👍s) need.

**Comment by TkDodo** [comment]
@AndreSilva1993 I think the solution could be simpler. Unless I'm mistaken, if you just want to know if something has been selected that hasn't started fetching yet because it's debounced, all you'd need is to compare selections:

```
const [selected, setSelected] = useState('')
const debouncedSelected = useDebounce(selected, 250)
const { data } = useQuery({
  queryKey: ['key', debouncedSelected],
  queryFn: () => fetchData(debouncedSelected)
})

const isTransitioning = selected !== debouncedSelected
```

now you can check for `isTransitioning` to show a loading spinner instead of data for item A (debouncedSelected), because you've selected itemB/C/D (selected) instead.

**Comment by TkDodo** [comment]
@vincerubinetti I get that this a feature requested by many, but there are always tradeoffs:
- bundle size vs. features
- api surface vs. simplicity

react-query does not have the smallest bundle size and not the most simplistic api surface. The stance is that features that can be implemented in user-land with minimal effort will not make it into the library. Interestingly, from the things you've mentioned: react-query doesn't do anything for "pagination". The pagination uses a feature that was meant to mimic suspense/transitions (keepPreviousData). Unless you're thinking of inifinte queries, which is something that isn't easily doable in user-land with raw useQuery.

Now for debouncing, the problems are:

- we'd need to implement debouncing or add an external dependency. right now, we have zero dependencies other than the query-core, which is our own code. `lodash/debounce` adds another kilobyte, about 10% of the total lib size. I know bundle-size is an argument that I personally dislike, too, but it's something that we sadly have to keep in mind.
- I haven't thought this through, but I think internal debounce is not easily compatible with how react-query works. consider the example from right above:

```
const [selected, setSelected] = useState('')
const { data } = useQuery({
  queryKey: ['key', setSelected],
  queryFn: () => fetchData(setSelected),
  // hypothetical api, doesn't exist
  debounce: 250
})
```

now, the queryKey would change with every input, but we wouldn't want to start observing that key (switching to that queryObserver) before debounce time. That also means that all other options changes should _not_ be applied, because they "belong to" the new selection key.
So I think this opens up a can of worms for no real gain. Lots of complexity and edge cases within the library. Basically, a key change doesn't / shouldn't do anything if debounce is specified. That is hard to justify and explain.

**Comment by AndreSilva1993** [comment]
> @AndreSilva1993 I think the solution could be simpler. Unless I'm mistaken, if you just want to know if something has been selected that hasn't started fetching yet because it's debounced, all you'd need is to compare selections:
> 
> ```
> const [selected, setSelected] = useState('')
> const debouncedSelected = useDebounce(selected, 250)
> const { data } = useQuery({
>   queryKey: ['key', debouncedSelected],
>   queryFn: () => fetchData(debouncedSelected)
> })
> 
> const isTransitioning = selected !== debouncedSelected
> ```
> 
> now you can check for `isTransitioning` to show a loading spinner instead of data for item A (debouncedSelected), because you've selected itemB/C/D (selected) instead.

Ok, I was definitely over-complicating things. I think your example more than suffices for what I'm trying to accomplish.

**Comment by AndreSilva1993** [comment]
@TkDodo Only caveat I've not realised with the solution you provided (should be a question of preference and will definitely ponder on this) is that even if react-query has already fetched the results for a specific item, I'll always display the loading spinner for the amount of time I'm debouncing items whereas, if the fetch is the one being debounced but the queryKey changes immediately, I get the cached results immediately.

**Comment by vincerubinetti** [comment]
@TkDodo Thank you for your comment. I think it's important for a maintainer to weigh in and explain why something is  "wont-fix". Also maybe the examples I gave were bad, but as you alluded to, RQ has a relatively complex api with a lot of options and features, which was my overall point.

From your explanation, I'm assuming that the "debouncing the serialized key" suggestion above falls prey to the same disproportionate increase in complexity you mentioned? Hopefully that idea has been adequately considered by the team too.

**Comment by SrBrahma** [comment]
I've been using [use-debounce](https://www.npmjs.com/package/use-debounce) since I commented here 2 weeks ago and it works great (lodash one didn't seem too good, maybe as it's not a proper tool for react?).

Still, I would like it if react-query had plugins to be optionally installed, so they wouldn't add additional weight for common users but only for those who opt in for that. I don't know much about this topic so I can't go further on it, but this is my conclusion about this.

**Comment by TkDodo** [comment]
> From your explanation, I'm assuming that the "debouncing the serialized key" suggestion above falls prey to the same disproportionate increase in complexity you mentioned? Hopefully that idea has been adequately considered by the team too.

Pretty much, yes. You would have to debounce the whole options:

```
useQuery({
  queryKey: ['key', input],
  queryFn: () => fetchData(input),
  debounce: 500,
  refetchInterval: input === 0 ? 0 : 5000
})
```

by "only" debouncing the key, we would put a refetchInterval on the input with value 0 if we transition from 0 to 1 for example. We would have to debounce _everything_, which is literally what debouncing the input in userland is.

> I'll always display the loading spinner for the amount of time I'm debouncing items whereas, if the fetch is the one being debounced but the queryKey changes immediately, I get the cached results straight away.

Yeah that's true, it's a general tradeoff when debouncing the queryKey. But it's the same issue as written above - I don't think we can easily just debounce the key or the queryFn, we need to debounce _everything_.

Don't want to sunshine the tradeoff, but I think it usually doesn't matter much to have a bunch of millis delay in rendering, and it might even be a good thing. If I'm typing `hello world` and I have results in the cache for `hello` and `hello w` and hello `wo`, I would get 3 intermediate renders with data I'm not really interested at. You might also need to go into concurrent features then to make those renders smooth.

> Edit: Or maybe at least add those solutions to the Docs so they don't appear hacky to bosses?

absolutely, PRs do the docs are always welcome ❤️

**Comment by jcdb95** [comment]
If you are here looking for some Vue answer, I used the [watchDebounced](https://vueuse.org/shared/watchdebounced/#watchdebounced) from [vueUse](https://vueuse.org/) and it looks something like this:

```
const rates = reactive(
  useQuery(
    ['rates', someDynamicKey],
     () => {
      return api.someApiCall()
    },
    {
      retry: 0,
      refetchOnWindowFocus: false,
      enabled: false
    }
  )
);

watchDebounced(
  rateData,
  () => {
    rates.refetch();
  },
  { debounce: 500, deep: true, immediate: true }
);
```

Hope it helps!

**Comment by julioflima** [comment]
> I've been using [use-debounce](https://www.npmjs.com/package/use-debounce) since I commented here 2 weeks ago and it works great (lodash one didn't seem too good, maybe as it's not a proper tool for react?).
> 
> Still, I would like it if react-query had plugins to be optionally installed, so they wouldn't add additional weight for common users but only for those who opt-in for that. I don't know much about this topic so I can't go further on it, but this is my conclusion about this.
> 
> Edit: Or maybe at least add those solutions to the Docs so they don't appear hacky to bosses?

Best solution ever, no one posted the safety lib to do the process. Of course this should be in the production by default.

Simple case super recurrent:
You gonna receive a parameter from the url to mutate when the user load the time for the first time. In the React 18, the useEffect is fetched two times, and this couldn't be done in the ServerComponents.

**Comment by thefaded** [comment]
You don't really need any of React hooks to handle debounce, here's an example with old version of `axios`.
```typescript
const debouncedFetcher = (url: string, params?: any, signal?: AbortSignal) => {
  const CancelToken = axios.CancelToken
  const source = CancelToken.source()
  /** Wrap your fetcher with Promise and resolve on timeout */
  const promise = new Promise((resolve) => setTimeout(resolve, 300)).then(() => fetcher(url!, params, source.token))

  signal?.addEventListener("abort", () => {
    source.cancel("Query was cancelled by React Query")
  })

  return promise
}
```

**Comment by hitsthings** [comment]
I'm here because I need a debounced mutate in a way it doesn't seem anyone else does. I'm using a mutation to save changes to a prose document. Since the user can be typing very fast, I only update once per second.

I'm looking for a debounced mutation that interacts well with optimistic queries.  So the version above at  https://github.com/TanStack/query/issues/293#issuecomment-1238760706 doesn't suffice.

I'd like to be able to do:
```js
const mutate = useMutation(somethingExpensive, {
    onMutate: ...set queries to optimistic value...,
    onError: ...revert queries...,
    onSettled: ...invalidate queries so they refetch...,
    debounce: 1000
})
```
And have `onMutate` called immediately so that the optimistic value is shown to the user without a delay.

The snippet linked wouldn't show the optimistic value until 1s later.

This is also a non-trivial feature to implement because if you implement it _within_ your query function, the mutation calls can pile up - do you really want that? There's no good response to the intermediate queries that never actually ran. It's not an error, but it's also not success.

And if you implement it outside of useMutation, you've got to reimplement/mess with your own onMutate and onSettled for optimistic queries.

Here's roughly what I have now. Call it like:

```js
const mutate = useMyModifiedMutation(fn, { debounce: 1000, onMutate: ... onSettled: ... })
```

Implemented like (not exactly - paraphrased, hopefully still works after the paraphrasing):
```js
function useMyModifiedMutation(fn, opts) {
    const queryClient = useQueryClient()
    const _onMutate = options.onMutate
    let mutationFn, onMutate, onSettled = opts.onSettled
    const fnRef = useRef(fn)
    fnRef.current = fn
    const [debouncedFn, onSettledDebounced = useMemo(() => opts.debounce
        ? [
            debounceP(args => fnRef.current(args), opts.debounce),
            onSettled && debounce(onSettled, 10) // otherwise you get all the onSettled's happening at once
        ]
        : [],
        [opts.debounce]
    )
    if (debouncedFn) {
        let debouncedContext
        mutationFn = _onMutate
            ? async (vars) => {
                const ret = debouncedFn(vars)
                // debouncedContext is always reset to the newest version of itself
                 _onMutate && (debouncedContext = await _onMutate(vars))
                return ret
            }
            : debouncedFn
        onMutate = _onMutate && (vars) => { // no-op, this has already happened.
            //just need to return the context we've been using
            return debouncedContext
        }
        onSettled = onSettledDebounced
    } else {
        onMutate = _onMutate
        mutationFn = fn
    }
    return useMutation({
        ...opts,
        onMutate,
        mutationFn
    })
}
```

where debounceP is

```ts
const getResolvable = <T>() => {
    let res : (p:Promise<T>) => void, p = new Promise<T>((resolve) => {
        res = resolve
    })
    return [res!, p] as const
}
const debounceP = <Args extends any[], Ret>(fn: (...args:Args) => Promise<Ret>, ms: number) : ((...args:Args) => Promise<Ret>) => {
    let handle: number | NodeJS.Timeout
    let [res, p] : readonly [null|((p:Promise<Ret>) => void), null|Promise<Ret>] = [null, null]
    const debounced = (...args: Args) => {
        if (res == null) {
            ([res, p] = getResolvable<Ret>() )
        }
        clearTimeout(handle)
        handle = setTimeout(() => {
            res && res(fn(...args))
            res = p = null
        }, ms)
        return p!
    }
    return debounced
}
```

And `debounce` is similar with out the Promise stuff:

```ts
const debounce = <Args extends any[]>(fn: (...args: Args) => void, ms: number) => {
    let handle: number | NodeJS.Timeout
    const debounced = (...args: Args) => {
        clearTimeout(handle)
        handle = setTimeout(() => fn(...args), ms)
    }
    return debounced
}
```

The big gotcha to this implementation is that intermediate mutations are thrown away! So you have to be sure that the latest mutation always includes previous ones as well (which is true in my case, based on the nature of text editing).

Hope this helps someone else!

**Comment by jjorissen52** [comment]
> For anyone who is interested in debouncing a mutation, I hacked together this hook. It returns a mutation with the `debouncedMutate` function which accepts `debounceMs` as an option. Apologies for no typescript.
> 
> ```js
> const useDebouncedMutation = (mutationFn, options) => {
>   const mutation = useMutation(mutationFn, options);
>   const [isDebouncing, setIsDebouncing] = useState(false);
>   let timer;
> 
>   const debouncedMutate = (variables, { debounceMs, ...options }) => {
>     clearTimeout(timer);
>     setIsDebouncing(true);
>     timer = setTimeout(() => {
>       mutation.mutate(variables, options);
>       setIsDebouncing(false);
>     }, debounceMs);
>   };
> 
>   return { isDebouncing, debouncedMutate, ...mutation };
> };
> ```
> 
> Example use:
> 
> ```js
> const channelExists = useDebouncedMutation(api.doesChannelExist);
> channelExists.debouncedMutate(channelName, {
>   debounceMs: 500,
> });
> ```
> 
> I'm using this hook in a chat app to check if a user inputted channel name already exists.

This implementation actually does not debounce. Though the  debouncedMutate closes over the context containing `timer`, the block `let timer` will execute every time the component re-renders, which means clearTimeout will always receive a `null` and never clear a timeout. Here is a modified version that will actually debounce:

```typescript
type DebouncedMutate<TData = unknown, TError = DefaultError, TVariables = void, TContext = unknown> = (
  variables: TVariables,
  { debounceMs, ...options }: UseMutationOptions<TData, TError, TVariables, TContext> & { debounceMs: number }
) => void
type UseDebouncedMutationReturn<TData = unknown, TError = DefaultError, TVariables = void, TContext = unknown> = Omit<
  UseMutationResult<TData, TError, TVariables, TContext>,
  'data' | 'mutate'
> & { debouncedMutate: DebouncedMutate<TData, TError, TVariables, TContext> }

export function useDebouncedMutation<TData = unknown, TError = DefaultError, TVariables = void, TContext = unknown>(
  options: UseMutationOptions<TData, TError, TVariables, TContext>
): UseDebouncedMutationReturn<TData, TError, TVariables, TContext> {
  const { mutate, ...mutation } = useMutation<TData, TError, TVariables, TContext>(options)
  const timer = useRef<NodeJS.Timeout>()
  const debouncedMutate: DebouncedMutate<TData, TError, TVariables, TContext> = (
    variables,
    { debounceMs, ...options }
  ) => {
    clearTimeout(timer.current)
    timer.current = setTimeout(() => {
      mutate(variables, options)
    }, debounceMs)
  }
  return { debouncedMutate, ...mutation }
}
```

**Comment by nickgrealy** [comment]
> You don't really need any of React hooks to handle debounce, here's an example with old version of `axios`.
> 
> ```ts
> const debouncedFetcher = (url: string, params?: any, signal?: AbortSignal) => {
>   const CancelToken = axios.CancelToken
>   const source = CancelToken.source()
>   /** Wrap your fetcher with Promise and resolve on timeout */
>   const promise = new Promise((resolve) => setTimeout(resolve, 300)).then(() => fetcher(url!, params, source.token))
> 
>   signal?.addEventListener("abort", () => {
>     source.cancel("Query was cancelled by React Query")
>   })
> 
>   return promise
> }
> ```

I like this answer from @thefaded ... I don't see why we can't just manage debouncing inline, and let `react-query` abort old requests.

e.g.
```javascript
const sleep = (ms = 1000) => new Promise(resolve => setTimeout(resolve, ms))

const someQuery = useQuery({
    queryKey: ['someKey'],
    queryFn: async ({ signal }) => {
        await sleep(500)
        if (!signal?.aborted) {
            return (await axios.get(`/api/someRequest`, { signal })).data
        }
    }
});
```

**Comment by bobbjedi** [comment]
```
const tanstackDebouncer = <T>({ fn, delaySeconds }: {
  fn: () => Promise<T>;
  delaySeconds: number;
}) => {

  let timerTimeOut: ReturnType<typeof setTimeout> = null

  const list: Array<(value: unknown) => void> = []

  const trigger = () => {

    timerTimeOut && clearTimeout(timerTimeOut)

    timerTimeOut = setTimeout(async () => {
      let resp: T = null
      try {
        resp = await fn()
      } catch (error) {
        console.error('Error trigger tanstack debouncer:', error)
      }
      list.forEach(r => r(resp))
      list.length = 0
    }, delaySeconds * 1000)
  }

  return () => {
    trigger()
    return new Promise(r => {
      list.push(r)
    })
  }
}

// Example use:
import { useQuery } from '@tanstack/vue-query';
import { Ref, unref } from 'vue';

export default ({ counter }: { counter: Ref<number> }) => {
  const debouncer = tanstakDebouncer<string[]>({
    fn: () => {
      return api.putCounter(unref(counter)).then((data) => data);
    },
    delaySeconds: 1,
  });

  return useQuery({
    queryKey: ['counter'],
    queryFn: () => {
      return debouncer();
    },
  });
};
```

**Comment by MartinCura** [comment]
Hi all, some integrated debouncing would be sweet, this is how i solved it so that cached results are shown immediately (e.g. user types "A", types "B", then goes back to "A"), basically having my own set of used query keys and using that to know whether to use the debounced or the immediate input value:

```tsx
// simplified from my code
const [inputValue, setInputValue] = useState<string>("")
const [searchQueriesSet, setSearchQueriesSet] = useState<Set<string>>(new Set())

const fetchSuggestions = useCallback(async (search: string) => {
  const response = await fetch(`/api/search?query=${search}`)
  const suggestions = searchResultsSchema.parse(await response.json())
  // Workaround for react-query not having debouncing, c.f. https://github.com/TanStack/query/issues/293
  setSearchQueriesSet((prev) => new Set([...prev, search]))
  return suggestions
}, [])

const debouncedInputValue = useDebounce(inputValue, 2000)
const searchQuery = searchQueriesSet.has(inputValue) ? inputValue : debouncedInputValue
const searchEnabled = searchQuery.length >= 8

const { data: suggestions, isFetching } = useQuery({
  queryKey: ["geo", "address-search", searchQuery],
  queryFn: () => fetchSuggestions(searchQuery),
  enabled: searchEnabled,
  staleTime: Infinity,
  gcTime: Infinity,
})
```

In case it's of help to anyone (or you have better suggestions, but i think this is fine for me).
